pulser-simulation/pulser_simulation/qutip_result.py

```diff
@@ -0,0 +1,184 @@
+    """
+
+    state: qutip.Qobj
+    matching_meas_basis: bool
+
+    @property
+    def sampling_errors(self) -> dict[str, float]:
+        """The sampling error associated to each bitstring probability."""
+        return {bitstr: 0.0 for bitstr in self.probability_dist}
+
+    @property
+    def _dim(self) -> int:
+        return int(self.state.dims[0][0])
+
+    @property
+    def _basis_name(self) -> str:
+        if self._dim > 2:
+            return "all"
+        if self.meas_basis == "XY":
+            return "XY"
+        if not self.matching_meas_basis:
+            return (
+                "digital"
+                if self.meas_basis == "ground-rydberg"
+                else "ground-rydberg"
+            )
+        return self.meas_basis
+
+    def _weights(self) -> np.ndarray:
+        n = self._size
+        if self.state.type != "ket":
+            probs = np.abs(self.state.diag())
+        else:
+            probs = (np.abs(self.state.full()) ** 2).flatten()
+
+        if self._dim == 2:
+            if self.matching_meas_basis:
+                # State vector ordered with r first for 'ground_rydberg'
+                # e.g. n=2: [rr, rg, gr, gg] -> [11, 10, 01, 00]
+                # Invert the order ->  [00, 01, 10, 11] correspondence
+                # The same applies in XY mode, which is ordered with u first
+                weights = (
+                    probs if self.meas_basis == "digital" else probs[::-1]
+                )
+            else:
+                # Only 000...000 is measured
+                weights = np.zeros(probs.size)
+                weights[0] = 1.0
+
+        elif self._dim == 3:
+            if self.meas_basis == "ground-rydberg":
+                one_state = 0  # 1 = |r>
+                ex_one = slice(1, 3)
+            elif self.meas_basis == "digital":
+                one_state = 2  # 1 = |h>
+                ex_one = slice(0, 2)
+            probs = probs.reshape([3] * n)
+            weights = np.zeros(2**n)
+            for dec_val in range(2**n):
+                ind: list[Union[int, slice]] = []
+                for v in np.binary_repr(dec_val, width=n):
+                    if v == "0":
+                        ind.append(ex_one)
+                    else:
+                        ind.append(one_state)
+                # Eg: 'digital' basis : |1> = index2, |0> = index0, 1 = 0:2
+                # p_11010 = sum(probs[2, 2, 0:2, 2, 0:2])
+                # We sum all probabilites that correspond to measuring
+                # 11010, namely hhghg, hhrhg, hhghr, hhrhr
+                weights[dec_val] = np.sum(probs[tuple(ind)])
+        else:
+            raise NotImplementedError(
+                "Cannot sample system with single-atom state vectors of "
+                "dimension > 3."
+            )
+        # Takes care of numerical artefacts in case sum(weights) != 1
+        return cast(np.ndarray, weights / sum(weights))
+
+    def get_state(
```

Comment:
This is likely the function I was mentioning could be added to the base interface.

Reply:
Maybe I'm looking at things the wrong way, but why we want a method that is particular to some result types to be defined in the base interface?

Reply:
Because getting the state should be a common operation for all the backends, imo.

Reply:
But then how would that work for QPUs?

Reply:
Either we create two classes from the parent `Results` or, for the QPU, we simply return a `NotImplementedError`, both are fine for me (the first one is likely more elegant but a bit more cumbersome to implement).